---
title: Detecting CD8 TPEX with ProjecTILs
date: "`r Sys.Date()`"
author: "P. Gueguen, M. Andreatta and S. Carmona"
output:
  rmdformats::readthedown:
    self-contained: true
    highlight: haddock
    thumbnails: false
    css: styles.css
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file, encoding=encoding,
  output_file=file.path(dirname(input_file), out_dir, 'Detecting_TPEX.html'))})
---

In this tutorial we highlight ProjecTILs ability to detect CD8+ ["TPEX" (Progenitor of exhausted)](https://www.nature.com/articles/s41590-022-01219-w), a rare population of memory CD8+ T cells that sustain CD8 T cell immunity in the context of chronic viral infection and cancer. These CD8 T cells are typically defined by co-expression of TCF1, PD-1 and TOX. They are of primary importance in cancer therapy, as they are thought to renew the pool of terminally exhausted CD8 T cells in response to PD-1/PD-L1 blockade [Siddiqui et al.](https://pubmed.ncbi.nlm.nih.gov/30635237/)

![](images/image-1262540734.png){width="600"}

Tumor T cell differentiation model, going through an intermediate state of progenitor exhausted (TPEX). Figure from [Andreatta et al. 2021](https://www.nature.com/articles/s41467-021-23324-4/figures/9).

ProjecTILs classifies cells by projecting them onto a reference map. Here we will use our reference map of human tumor-infiltrating CD8 T cells. You can find more information about this map [here](https://github.com/carmonalab/CD8_human_TIL_atlas_construction).

<details>

<summary style="display:list-item">

Why using ProjecTILs to classify cell subtypes?

</summary>

-   Obtain consistent cell annotations across datasets

-   Classify cells into reference subtypes, irrespective of activation of transient gene programs, such as cell cycle.

-   Avoid use of subjective, dataset-specific parameters, such as the selection of highly variable genes, number of clusters, etc.

-   Projection is robust to batch effects, single-cell technologies, sequencing depth (for more information, please read ProjecTILs paper - [Andreatta et al. 2021](https://www.nature.com/articles/s41467-021-23324-4)).

</details>

<details>

<summary style="display:list-item">

What happens if the input data contains cells that do not match the the reference map cell type?

</summary>

If the input data contain cell types not included in the reference map(eg. CD4 T cells if the reference is for CD8 T cells) they automatically get filtered-out.

By default, both `Run.ProjecTILs()` and `ProjecTILs.classifier()` have the parameter `filter.cells` set as `TRUE`. This means that cells out of reference will be filtered-out using the built-in [`scGate`](https://github.com/carmonalab/scGate) model. This model is stored in the slot misc of the reference [`Seurat`](https://github.com/satijalab/seurat) object: `ref@misc$scGate`. You can custom this filtering by amending this slot using [`scGate`](https://github.com/carmonalab/scGate) grammar.

</details>

</details>

```{r, include=FALSE, warning=FALSE, message=FALSE}
renv::restore()
library(Seurat)
library(tidyverse)
library(scGate)
library(STACAS)
library(data.table)
library(GEOquery)
library(patchwork)
library(ProjecTILs)
library(gridExtra)
library(SignatuR)
library(ggrepel)
library(scales)
library(UCell)
library(pheatmap)
library(EnhancedVolcano)
library(plotly)
#options(future.globals.maxSize= 5000*1024^2)
```

# Human CD8 TIL reference

First, let's have a look at the reference map.

```{r, message = F, warning=FALSE}
# Load the reference
options(timeout = max(900, getOption("timeout")))
#download.file("https://figshare.com/ndownloader/files/38921366", destfile = "CD8T_human_ref_v1.rds")
ref.cd8 <- load.reference.map("CD8T_human_ref_v1.rds")

# Setup colors
mycols <- ref.cd8@misc$atlas.palette

# DimPlot
DimPlot(ref.cd8,  group.by = 'functional.cluster', label = T, repel = T, cols = mycols) + theme(aspect.ratio = 1)
```

Here are the different T cell subsets defined in the map:

-   [**CD8.NaiveLike:**]{style="color:#b7d2e0"} Antigen-naive T cells

-   [**CD8.CM:**]{style="color:#da6f6f"} Central Memory T cells

-   [**CD8.EM:**]{style="color:#72b28a"} Effector Memory

-   [**CD8.TEMRA:**]{style="color:#e5bfaf"} Effector Memory cells re-expressing CD45RA. Sometimes called Short Lived Effectors (SLEC), or Cytotoxic effectors

-   [**CD8.TPEX:**]{style="color:#aca6e0"} Progenitors exhausted T cells

-   [**CD8.TEX:**]{style="color:#f5d39f"} Exhausted T cells

-   [**CD8.MAIT:**]{style="color:#fdbfd4"} Mucosal-associated invariant T cells, innate-like T cells defined by their semi-invariant αβ T cell receptor

Let's check Differentially Expressed Genes (DEGs) between subtypes and verify expected marker genes:

```{r, message = F, warning=FALSE, fig.height=10, fig.width=7}
# Compute DEGs
DefaultAssay(ref.cd8) <- "RNA"
ref.cd8 <- NormalizeData(ref.cd8)
markers <- FindAllMarkers(object = ref.cd8, only.pos = TRUE, assay = "RNA")

# Remove TCR genes
tcr.genes <- SignatuR::GetSignature(SignatuR$Hs$Compartments$TCR)
markers <- markers %>% filter(!gene %in% unname(tcr.genes))
markers %>% group_by(cluster) %>% top_n(n = 3, wt = avg_log2FC) -> top3

# Plot heatmap
VlnPlot(ref.cd8, assay = "RNA", features  = top3$gene, cols = mycols, stack = T, flip = T, fill.by = "ident") + NoLegend()
```

Here are some marker genes to help manually identify TPEX:

[Positive markers:]{.underline} TCF7, CD200, CRTAM, GNG4, TOX, LEF1, CCR7, CXCL13, XCL1, XCL2

[Negative markers:]{.underline} GZMB, NKG7, PRF1, HAVCR2, CCL5, GZMA

Let's display 6 positive and 6 negative TPEX markers, that are especially useful to distinguish TPEX from closely related TEX.

```{r fig.width=10, fig.height=15}
DefaultAssay(ref.cd8) <- 'RNA'
FeaturePlot(ref.cd8, features = c('TCF7','XCL1','XCL2',"CXCL13","TOX","CRTAM","GZMB","NKG7","CCL5","HAVCR2","PRF1","GZMA"), ncol = 3, pt.size = 0.2, order = T, cols = pals::coolwarm()) & NoLegend()
```

Despite their great releveance, TPEX are often missed in tumor scRNA-seq studies.

# Detecting TPEX in [Gueguen et al. 2021](https://www.science.org/doi/10.1126/sciimmunol.abd5778)

## Setup data

```{r, message=F, warning=F}
#download.file("https://figshare.com/ndownloader/files/39082049", destfile = "gueguen.cd3.Rds")
gueguen.cd3 <- readRDS("gueguen.cd3.Rds")
gueguen.cd3$seurat_clusters <- Idents(gueguen.cd3)
```

## Projection

Thanks to automatic scGate filtering, only the CD8 clusters (upper part of the UMAP) are annotated.

```{r, message=F, warning=F, fig.height=5, fig.width=10}
# Projection
DefaultAssay(gueguen.cd3) <- "RNA"
gueguen.cd3 <- ProjecTILs.classifier(gueguen.cd3, ref = ref.cd8, filter.cells = T, split.by = 'orig.ident', ncores = 6)
table(gueguen.cd3$functional.cluster)
DimPlot(gueguen.cd3, order = T,  label = T, repel = T) 
```

```{r}
DimPlot(gueguen.cd3, group.by = 'functional.cluster', order = T, cols = mycols, label = T, repel = T)
```

We can check specific TPEX genes to confirm that identities match.

```{r fig.width=10, fig.height=4}
DefaultAssay(gueguen.cd3) <- 'RNA'
FeaturePlot(gueguen.cd3, features = c('XCL1','XCL2'), ncol = 2, pt.size = 0.5, order = T, cols = pals::coolwarm()) & NoLegend()
```

```{r, fig.height=8, fig.width=15}
# Radar plots
p <- plot.states.radar(ref.cd8, query = gueguen.cd3, min.cells = 10, genes4radar = c('LEF1', "TCF7", "CCR7", "GZMK", "FGFBP2",'FCGR3A','ZNF683','ITGAE', "CRTAM", "CD200",'GNG4', "HAVCR2", "TOX", "ENTPD1", 'TYROBP','KIR2DL1'), return = T) 
wrap_plots(p) + theme_bw()
```

We can see that the previously homogeneous cluster CD8-LAYN is actually composed of two subsets: CD8.TEX and CD8.TPEX.

<details>

<summary style="display:list-item">

How to assess quality of mapping?

</summary>

The first thing we recommend to do is verifying the expression of expected marker genes. In addition, ProjecTILs provides multiple tools to help the researcher decide if the projection and classification are accurate (e.g. <https://carmonalab.github.io/ProjecTILs_CaseStudies/novelstate.html>).

</details>

# Detecting TPEX in [Yost et al. 2019](https://www.nature.com/articles/s41591-019-0522-3)

## Setup data

```{r}
# Load data
#download.file("https://figshare.com/ndownloader/files/39109277", destfile = "Yost.cd3.Rds")
Yost.cd3 <- readRDS("Yost.cd3.Rds")

# Normalize data
Yost.cd3 <- NormalizeData(Yost.cd3)
Yost.cd3 <- ScaleData(Yost.cd3)

# DimPlots
DimPlot(Yost.cd3, reduction = 'umap', group.by = 'cluster', label = T)
DimPlot(Yost.cd3, reduction = 'umap', group.by = 'patient', label = T, repel = T)
```

## Projection

As this dataset is a mix between CD4 and CD8 T cells, we will keep the parameter `filter.cells` as TRUE to keep only CD8+ T cells.

```{r, message=F, warning=F}
DefaultAssay(Yost.cd3) <- "RNA"
Yost.cd3 <- ProjecTILs.classifier(Yost.cd3, ref = ref.cd8, filter.cells = T, split.by = 'patient', ncores = 6)
table(Yost.cd3$functional.cluster)

DimPlot(Yost.cd3, group.by = 'functional.cluster', order = T, cols = mycols, label = T, repel = T)
```

Here again we detect TPEX. Let's check globally how the expression profiles of marker genes look.

```{r, fig.height=8, fig.width=15}
# Radar plots
p <- plot.states.radar(ref.cd8, query = Yost.cd3, min.cells = 10, genes4radar = c('LEF1', "TCF7", "CCR7", "GZMK", "FGFBP2",'FCGR3A','ZNF683','ITGAE', "CRTAM", "CD200",'GNG4', "HAVCR2", "TOX", "ENTPD1", 'TYROBP','KIR2DL1'), return = T) 
wrap_plots(p) + theme_bw()
```

We can see that predicted CD8.TPEX displey consistent marker gene profiles. In the authors' original UMAP space, however, TPEX are found scattered. This is because activation signals and other confounding factors were contributing to defining the UMAP space. Reference-based annotation uncovers cell type signals masked by the activation program. If you are interested in recovering cell type identities hidden by transient cell states, you can read more in the [corresponding tutorial](https://carmonalab.github.io/human_CD8_TIL_CaseStudies/Transient_gene_programs.html).

<details>

<summary style="display:list-item">

Session Info

</summary>

```{r}
sessionInfo()
```

</details>
